### ISSUE TITLE
OWD project: update existing events

### ISSUE THREAD

**Issue by foolip** [openwebdocs/project#61]
The way that events are documented on MDN and represented in BCD is not great, with different facets of events being treated as separate, and no clear entry point to learn about an event. Previous discussion on this is in https://github.com/mdn/browser-compat-data/issues/7545 and https://github.com/mdn/browser-compat-data/issues/12290.

To summarize the three main facets in terms of BCD entries / MDN pages that often co-exist:
- Event interfaces, such as [`ClipboardEvent`](https://developer.mozilla.org/en-US/docs/Web/API/ClipboardEvent)
- \*_event entries/pages such as [`cut_event`](https://developer.mozilla.org/en-US/docs/Web/API/Element/cut_event) (in this case there are [three](https://developer.mozilla.org/en-US/docs/Web/API/Document/cut_event) [more](https://developer.mozilla.org/en-US/docs/Web/API/SVGGraphicsElement/cut_event) [pages](https://developer.mozilla.org/en-US/docs/Web/API/Window/cut_event))
- on* event handler attributes, such as [`oncut`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/oncut) (these are often on many interfaces, often via the `GlobalEventHandlers` mixins)

Suggested work:
- Avoiding the duplication between \*_event and on\* entries and pages, and have a clear main page for each event such as "cut". While `element.addEventListener('cut', handler)` and `element.oncut = handler` aren't exactly the same and can have different browser support, I think we should think of this as one feature, which could have partial support.
- Get rid of the last mixins: `GlobalEventHandlers` and `WindowEventHandlers` (https://github.com/mdn/browser-compat-data/issues/12290)
- Ensure the correct linkage between events and their interface, so that one can always find the right interface for an event ("cut" &rarr; `ClipboardEvent` and `ClipboardEvent` &rarr; "cut" + "copy" + "paste")

**Comment by Elchi3** [comment]
Thanks for filing a project proposal for improvements of MDN's event docs! I agree event docs need work and the items you suggest are good issues to look into.

I started https://github.com/mdn/content/discussions/9098 to discuss your first suggested work item. If a specific proposal can reach consensus there, work could start.

I will look at your other suggested work items, too, but they seem not blocked by any prior discussion.

**Comment by ddbeck** [comment]
@Elchi3 and I chatted about this idea earlier today. I like this idea a lot, but, if it's accepted, I'd like to sugest a couple of work items so that this can fit in with other content organization, structure, and tooling work:

- Uniquely identifying this _kind_ of page, among others (e.g., a page type name and version or date or something—perhaps existing event pages are api-event-1.0 and these are api-event-2.0 or somesuch—that could be recorded in frontmatter to show an author's intent to follow this new convention)
- Writing a detailed specification for the page type (e.g., describing expected titles, sections, etc.) so that authors not directly involved in this effort can determine whether they're producing content that's congruent with it

**Comment by wbamberg** [comment]
While we are doing this (and also related to defining the event page type) I'd love to make the tables:

<img width="825" alt="Screen Shot 2021-09-20 at 11 25 43 AM" src="https://user-images.githubusercontent.com/432915/134054570-b3fc905c-dbac-46ae-b5ba-a4970f9f5ab4.png">

...not be hardcoded, but built out of metadata. Maybe webref  has or could have the data, even.

**Comment by foolip** [comment]
@wbamberg I like that idea. Unfortunately none of that is current machine readable in webref or possible to derive from specs  automatically AFAICT, as the bubbles/cancelable/interface bits are almost always spelled out using prose in specs.

The relevant event handler properties probably could be derived given just some basic hard-coded facts about things bubble, i.e. knowing that events bubble through `Element` nodes though a `Document` and finally a `Window`. Without that knowledge, it would be hard to tie api.Element.mouseover_event to api.Window.onmouseover, for example.

**Comment by saschanaz** [comment]
Data about events using more specific event interfaces (not just `Event`) are in @types/web generator (see "events" fields in [addedTypes.jsonc](https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/main/inputfiles/addedTypes.jsonc) and [overridingTypes.jsonc](https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/main/inputfiles/overridingTypes.jsonc)), but no bubbles/cancelable unfortunately.

**Comment by jpmedley** [comment]
I'm still confused about something. I've heard this proposal described as 'one event page', It sounds like event interfaces will still exist, unless I'm misreading this. I think the confusion might be around the meaning of 'event'. I always (or try to) refer either 'event interface', 'event callback' or 'event handler'. If the proposal is simply to merge callbacks and handlers to 'event callback/handler' or some such thing, I'm completely on board with that. 

The distinction 'event interface' and 'event callback/handler' is the one I'm most concerned about. This morning I was reviewing Rachel's drafts for `FontFaceSet`. It has three callback/handlers, but they're all passed the same event interface type. In this case we're lucky that these were implemented as part of the same spec. I'm more concerned about events that are shared by interfaces and specs. The `Event` interface is very old, but it could be passed to a callback/handler that was implemented yesterday and the BCD would need to reflect that.

**Comment by wbamberg** [comment]
@jpmedley , although I didn't file this I will try to explain my understanding.

>  It sounds like event interfaces will still exist

Yes.

> If the proposal is simply to merge callbacks and handlers to 'event callback/handler' or some such thing, I'm completely on board with that.

I think it is.

Consider, say, the `"click"` event. On MDN at the moment we have 2 pages describing this event:

* [`Element/click_event`](https://developer.mozilla.org/en-US/docs/Web/API/Element/click_event): this page describes the event as an abstraction really, inasmuch as there's no property or method called `click_event`. This page means "you can call `addEventListener("click", ...)` on this interface, and here's all the information about that".
* [`GlobalEventHandlers/onclick`](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onclick): this page describes a concrete property `onclick` that this interface has, that allows you to listen for click events.

The proposal is to combine these two pages.

There's also the object that get passed into the click event handler, which MDN says is a [`MouseEvent`](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent), and which MDN (and you) call an "event interface". As you say, this type could appear in the handlers for many different events, so it wouldn't make sense to try to fold it into the event pages. It's a completely separate thing.

**Comment by foolip** [comment]
Indeed, there's no suggestion to merge event interfaces into the event pages.

**Comment by jpmedley** [comment]
The other thing that had confused me was that I thought this decision had been made already, quite some time ago. I've been submitting APIs without the abstraction page for quite some time and had hired contractors do the same. No one ever complained.

**Comment by foolip** [comment]
Which is the abstraction page you've been skipping, foo_event or onfoo?

**Comment by foolip** [comment]
@jpmedley what do you do when there are lots of `onfoo` event handler properties?

@Elchi3 do you know who will be driving this issue this quarter? @queengooborg are continuing towards 100% complete API data in BCD, and have to update both kinds of entries. Deciding what happens in BCD sooner rather than later would be great.

**Comment by foolip** [comment]
@jpmedley I see. If we get rid of `GlobalEventHandlers` by just documenting these properties like any others, it would result in both duplication (somewhat OK) but also a fair amount of useless documentation for event that the browser actually never fires. An example of this is `document.onvolumechange`, which exists, but the "volumechange" event is only ever fired at media elements and doesn't bubble, so this property is useless for web developers.

**Comment by foolip** [comment]
Also, there are events that *still* don't have event handler properties even in the spec, like https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event

**Comment by foolip** [comment]
Depending on how events in BCD are restructured as part of this project, it may not be necessary to figure out which on* event handler properties are pointless, at least not up front. In particular, with the kind of approach suggested in https://github.com/mdn/browser-compat-data/issues/7545#issuecomment-859721893, the on* properties would likely be moved one by one, and what's left at the end is the stuff that should be removed.

**Comment by jpmedley** [comment]
Let me rephrase. I need a list. I regularly generate [boilerplates for missing pages](https://github.com/jpmedley/mdn-boilerplates) and [make them available to contributors](https://github.com/mdn/content/discussions/9897). I need to not generate boilerplates for these events.

**Comment by Elchi3** [comment]
I made some lists here: https://github.com/mdn/content/discussions/9098#discussioncomment-1588303 
However, it is time consuming to get all events from specs given they aren't in IDL or in some structure, see https://github.com/w3c/reffy/issues/772 for an issue that aims to extract event data.

**Comment by Elchi3** [comment]
> Suggested work:
> 
> * Avoiding the duplication between *_event and on* entries and pages, and have a clear main page for each event such as "cut". While `element.addEventListener('cut', handler)` and `element.oncut = handler` aren't exactly the same and can have different browser support, I think we should think of this as one feature, which could have partial support.

I propose I start work by focusing on this aspect.

**Comment by jpmedley** [comment]
This seems reasonable to me. If we separated these, the two pages would be nearly identical. It's extra work with little gain.

**Comment by dontcallmedom** [comment]
@foolip @Elchi3 when you get a chance, @tidoust and I have made good progress in collecting information about events from across specifications, with the current version of our work ready to be released as a package reviewable in https://raw.githubusercontent.com/w3c/webref/curated/ed/events.json with the guarantees described at https://github.com/w3c/webref/tree/curated/packages/events#guarantees

If you could review this and bring feedback in the webref repo on improvements either to the structure or the data itself, this would be useful in determining next steps for this effort.

**Comment by queengooborg** [comment]
Great news: GlobalEventHandlers has been demixed from BCD and https://github.com/mdn/browser-compat-data/issues/14578 has been completed!

As mentioned a few times above, the linked issue was the criteria for completion of this project, however do we also want to include adding missing events, using the wonderful work from @dontcallmedom and @tidoust, as a part of this project's scope?  Or do we want to create a new project for that?

**Comment by teoli2003** [comment]
\o/

I would rename this project "Demixing GEH" and create a new one for "Add missing events".

Not that there is practically no difference (same work), but sometimes it is good to mark progress and to keep projects smaller (read scoped).

Else we will never be able to celebrate.

🎉

**Comment by queengooborg** [comment]
Thanks @teoli2003!  I brought this up during the SC call and it was generally agreed we should create a new project for adding the missing events and mark this one as completed.  I will create the new project now and close this one!  🎉
